keps/sig-release/20200122-kubernetes-yearly-support-period.md

```diff
@@ -0,0 +1,318 @@
+- Establish an “LTS” release process in the sense perhaps known from other projects (e.g.: pick one release every year or two, give patch support on that release for multiple years).
+- Declare that Kubernetes’ dependencies must also have a similar support lifetime per release.
+- Align Kubernetes releases in time with those of its dependencies.
+
+## Proposal
+
+The proposal is to extend the Kubernetes support cadence from 9 months to 14 months.
+The 14 months proposed here is essentially 12 months of support and 2 months of upgrade period.
+NOTE: This follows and formalizes today’s practice of retaining patch branch CI for 6-7 weeks beyond the 9 months intended support phase as a failsafe option if any highly critical CVEs or blocking upgrade issues are observed by slower adopters of newer patch branches.
+
+A visual example today vs. proposed for an adopter of 1.18 in April 2020:
+
+![Visual example of upgrade cadences](./20200122-upgrade-cadence.png)
+
+## Design Details
+
+### Component changes
+
+* Adjust behavior in the kube-apiserver that assumes a component lifetime of ~9 months
+  * Lifetime of the certificate used for kube-apiserver loopback connections (#86552)
+* Expand supported kubelet/apiserver skew to 12 months divided by release cadence minus 1, i.e. currently (12/3)-1=3 releases.
+  * Currently, supported skew (2 versions) was chosen to allow the oldest kubelet to work against newest apiserver
+  * Implementation should include adding a test to cover the added skew variations, have it release informing, watch for any errors, fix them, promote the test to release blocking.  This is already a gap today and can be treated as a project need orthogonal to this KEP.
+* Expand supported kubectl/apiserver skew
+  * Currently supported skew (+/- 1 minor version) theoretically allows using an n-1 kubectl to work against all supported apiservers (n-2/n-1/n).
+  * Ideally, the latest kubectl would support speaking to all supported apiservers (+0/-3 minor versions).
+  * Implementation should include adding a test to cover the added skew variations, have it release informing, watch for any errors, fix them, promote the test to release blocking.  This is already a gap today and can be treated as a project need orthogonal to this KEP.
+
+### Test/release process changes
+
+* Staff patch releases for one more branch for ~3-4 more months
+  * Feedback from SIG Release received, see [SIG Release meeting discussion](https://docs.google.com/document/d/1Fu6HxXQu8wl6TwloGUEOXVzZ1rwZ72IAhglnaAMCPqA/edit#bookmark=id.mi11nk75iohl).
+* Maintain CI for one more branch for ~3-4 more months
+  * Modifying test job and infrastructure patterns may need deprecated on a slower schedule or backwards compatibility with older style tests in older branches may need maintained longer.  This is a problem already today (e.g.: bootstrap.py went out of support, but was still in use in release-1.14 branch CI during the end of that branch’s lifetime).
```

Comment:
possibly better to exclude bootstrap.py as an example and just explain that images and jobs for another branch should be maintained in test-infra.

Reply:
I'm not familiar with bootstrap.py, but that sounds like it's really a deprecation warning rather than an out of support thing?

@tpepper mentioned that bootstrap.py caused a problem with the 1.14 release, I think that's where this came from.

Happy to remove, but I'll need a hand with a suggestion for a concrete example.

Reply:
not familiar with what happened in 1.14, but boostrap.py should be fully independent of kubernetes versions. the actual test configs and the test images are version dependent though